compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -1925,6 +1925,26 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                     *impl_,
                 );
                 if let Some(assoc_ty) = assoc_ty {
+                    let (assoc_ident, def_scope) =
+                        tcx.adjust_ident_and_get_scope(assoc_ident, adt_def.did(), hir_ref_id);
+
+                    let kind = DefKind::AssocTy;
+                    if !assoc_ty.visibility(tcx).is_accessible_from(def_scope, tcx) {
+                        let kind = kind.descr(assoc_ty.def_id);
+                        let msg = format!("{kind} `{assoc_ident}` is private");
+                        let def_span = self
+                            .tcx()
+                            .hir()
+                            .span_if_local(assoc_ty.def_id)
+                            .unwrap_or_else(|| self.tcx().def_span(assoc_ty.def_id));
+                        tcx.sess
+                            .struct_span_err_with_code(span, &msg, rustc_errors::error_code!(E0624))
+                            .span_label(span, &format!("private {kind}"))
+                            .span_label(def_span, &format!("private {kind} defined here"))
+                            .emit();
+                    }
+                    tcx.check_stability(assoc_ty.def_id, Some(hir_ref_id), span, None);
+
                     let ty = tcx.type_of(assoc_ty.def_id);
                     return Ok((ty, DefKind::AssocTy, assoc_ty.def_id));
                 }
```

Comment:
This whole block repeats a lot of what we do below for trait associated types.
Could we extract it as an `lookup_associated_type` which would do the lookup/privacy/stability and return an `Option<DefId>`?